### ISSUE TITLE
GraphQL fails with internal server error due to missing "public schema token"

### ISSUE THREAD

**Issue by denisyilmaz** [craftcms/cms#7925]
### Description

After Updating our live server to Craft CMS 3.6.14 we experience a error with our GraphQL endpoint.
We only use a public schema (which does not need a token). Locally this is working fine. On the live server we get an internal server error when trying to access `example.com/api`:
```json
{
"name": "Internal Server Error",
"message": "Es ist ein interner Serverfehler aufgetreten.",
"code": 0,
"status": 500
}
```
in the logs i found this error log:

```bash
2021-05-12 13:29:45 [-][2][][error][Error] Error: Call to a member function getIsValid() on null 2021-05-12 13:29:45 [-][2][][error][Error] Error: Call to a member function getIsValid() on null in /var/www/example.com/vendor/craftcms/cms/src/controllers/GraphqlController.php:275
Stack trace:
#0 /var/www/example.com/vendor/craftcms/cms/src/controllers/GraphqlController.php(240): craft\controllers\GraphqlController->_publicToken(Object(craft\services\Gql))
#1 /var/www/example.com/vendor/craftcms/cms/src/controllers/GraphqlController.php(106): craft\controllers\GraphqlController->_schema(Object(craft\services\Gql))
#2 [internal function]: craft\controllers\GraphqlController->actionApi()
#3 /var/www/example.com/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#4 /var/www/example.com/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams(Array)
#5 /var/www/example.com/vendor/craftcms/cms/src/web/Controller.php(190): yii\base\Controller->runAction('api', Array)
#6 /var/www/example.com/vendor/yiisoft/yii2/base/Module.php(534): craft\web\Controller->runAction('api', Array)
#7 /var/www/example.com/vendor/craftcms/cms/src/web/Application.php(278): yii\base\Module->runAction('graphql/api', Array)
#8 /var/www/example.com/vendor/yiisoft/yii2/web/Application.php(104): craft\web\Application->runAction('graphql/api', Array)
#9 /var/www/example.com/vendor/craftcms/cms/src/web/Application.php(263): yii\web\Application->handleRequest(Object(craft\web\Request))
#10 /var/www/example.com/vendor/yiisoft/yii2/base/Application.php(392): craft\web\Application->handleRequest(Object(craft\web\Request))
#11 /var/www/example.com/web/index.php(22): yii\base\Application->run()
#12 {main}
2021-05-12 13:29:45 [-][2][][info][application] $_GET = [
    'q' => '/api'
]
```

for whatever reason it seems our server instance requires a token for the public schema.
I checked locally and if i try to add a token there i get the correct message, that there is no schema which needs a token:

![image](https://user-images.githubusercontent.com/4189607/117970488-0262ea00-b329-11eb-94d9-09298b6c70fe.png)
![image](https://user-images.githubusercontent.com/4189607/117970526-0b53bb80-b329-11eb-9a31-8be3cdcde77c.png)

but on the server it looks like this:

![image](https://user-images.githubusercontent.com/4189607/117970676-35a57900-b329-11eb-8d69-2faa1ce14042.png)

When testing with `Insomnia` on the live instance i only can access the schema with the token from above.
Without the token i get the internal server error described above

### Steps to reproduce

... cant reproduce locally, the error only appears on the live instance.

### Additional info

- Craft version: 3.6.14
- PHP version: 7.3.26
- Database driver & version: PostgreSQL 10.15
- Plugins & versions:
  - Hasura : dev-next
  - Imager : v2.4.0
  - Mailjet : 1.2.3
  - Neo : 2.9.7
  - Redactor : 2.8.7

**Comment by andris-sevcenko** [comment]
What version of Craft where you updating from?

**Comment by denisyilmaz** [comment]
We were updating from 3.6.10 to 3.6.14

```
  "require": {
    "aelvan/imager": "v2.4.0",
    // "bertoost/craft-mailjet": "1.2.2",
    // "craftcms/cms": "3.6.10",
    // "craftcms/redactor": "2.8.5",
    "bertoost/craft-mailjet": "1.2.3",
    "craftcms/cms": "3.6.14",
    "craftcms/redactor": "2.8.7",
    "jasonmccallister/hasura": "dev-next",
    // "spicyweb/craft-neo": "2.9.3",
    "spicyweb/craft-neo": "2.9.7",
    "vlucas/phpdotenv": "^3.4.0"
  },
  ```

**Comment by andris-sevcenko** [comment]
That's odd, to say the least. Any chance you can send over your DB backup and composer files to support@craftcms.com? If so, please reference this issue, so we know what's it about!

**Comment by denisyilmaz** [comment]
will do, thanks.

**Comment by huelabs** [comment]
I may have experienced a similar issue. I found "Public Schema" was missing in the GraphQL Schemas
tab. My development server did not have this issue.

After a ton of troubleshooting, I found my issue was my production server general config file having `'allowAdminChanges' => false`. Once I changed to true and deployed the change to production, Public Schema was back and I could query. Not ideal since you should never want allowAdminChanges to be true on production.

**Comment by denisyilmaz** [comment]
@huelabs  thanks for this. I just tested your workaround and voila, after setting `'allowAdminChanges' => true` and invalidating data caches it works for me! i was even able to switch back to `'allowAdminChanges' => false` afterwards and the schema remained. As this is a live webpage I can't do an update to see if the `public schema` will remain but for now its working.

**Comment by huelabs** [comment]
@brandonkelly, is this a known issue?

**Comment by brandonkelly** [comment]
If you were to do the same thing locally, do you see a change to the project config that could be deployed?

**Comment by andris-sevcenko** [comment]
@huelabs @denisyilmaz the schema should not have disappeared from the production server. Any chance both of you could send over the `storage/config-deltas` folder to support@craftcms.com and reference this issue?

**Comment by denisyilmaz** [comment]
sure, will do.

**Comment by andris-sevcenko** [comment]
Okay. I just pushed a fix for the next Craft release.

What happened here is that the link between the public token and public schema got severe. Likely because the public schemas were created in different environments or something. Either way, I'm 99% sure that at some point project config changes got deployed, where the UID of the public schema was changed.

This caused Craft to drop the old public schema (which caused the foreign key relation to be severed for the special `__PUBLIC__` token) and to create a new public schema, which remained detached from the public token.

Long story (and behind the scenes technical details) short, Craft now is able to restore this link, in the rare cases where the public schema is not attached to the public token, yet they both exist.
